Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -638,9 +660,29 @@ private Task ForceCacheFlushAsync(OperationContext context)
                 }).ThrowIfFailure();
         }
 
-        internal void ForceCacheFlush(OperationContext context)
+        internal bool ForceCacheFlush(OperationContext context, ContentLocationDatabaseCounters? counter = null, bool blocking = true)
         {
-            ForceCacheFlushAsync(context).GetAwaiter().GetResult();
+            // Ensure only one thread can start the flush
+            if (Interlocked.CompareExchange(ref _isFlushInProgress, 1, 0) == 1)
+            {
+                return false;

```

Comment:
>return false; [](start = 16, length = 13)

Is it ok to early out in the case where a blocking flush is requested?

Reply:
No it's not, I added a wait handle for this case. Please re-review

Reply:
No it's not, I added a wait handle for this case. Please re-review

---
In reply to: [333181654](https://github.com/microsoft/BuildXL/pull/1021#discussion_r333181654) [](ancestors = 333181654)

Reply:
At the very least, I think you should use ManualResetEventSlim instead of using a system event wait handle.

---
In reply to: [333220081](https://github.com/microsoft/BuildXL/pull/1021#discussion_r333220081) [](ancestors = 333220081,333181654)